### PR TITLE
10 get /api/users

### PR DESCRIPTION
**Pull request by Spyabo**

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -482,3 +483,28 @@ describe("DELETE /api/comments/:comment_id", () => {
...
+      });
+  });
+
+  it("200: each user has 3 properties", () => {
+    return request(app)
+      .get("/api/users")
+      .expect(200)
+      .then(({ body }) => {
+        expect(body.users.length).toBe(4);
+        expect(body.users[0]).toMatchObject({
```
Comment (itsaliabdallah) [inline __tests__/app.test.js:503]: would be better to iterate through and check each user has these properties than just check the first user.

**Review by itsaliabdallah — CHANGES_REQUESTED** [review]
Please implement changes needed for Ticket 7 and resubmit.

File: __tests__/app.test.js [hunk 2]
```diff
@@ -482,3 +484,30 @@ describe("DELETE /api/comments/:comment_id", () => {
...
+        users.forEach((user) => {
+          expect(user).toMatchObject({
+            username: expect.any(String),
+            name: expect.any(String),
+            avatar_url: expect.any(String),
+          });
+        });
+      });
+  });
+});
```
Comment (popatre) [inline __tests__/app.test.js:513]: user tests look good

File: controllers/comments.controller.js [hunk 3]
```diff
@@ -8,7 +8,17 @@ exports.postComment = async (req, resp, next) => {
...
+      if (data.code === "22P02") {
+        resp.status(400).send({ msg: "Invalid ID" });
+      }
+      if (data.code === "23503") {
+        const table = data.detail.split(" ").at(-1);
+        if (table === '"users".') {
+          resp.status(404).send({ msg: "User not found" });
+        }
+        resp.status(404).send({ msg: "Article not found" });
+      }
```
Comment (popatre) [inline controllers/comments.controller.js:21]: This seems to be to do with errors, so should be handled in your PSQL error controller, not this one

File: controllers/users.controller.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+const { fetchUsers } = require("../models/users.model");
+
+exports.getUsers = (req, res, next) => {
+  return fetchUsers()
+    .then((users) => {
+      res.status(200).send({ users });
+    })
+    .catch((err) => {
+      next(err);
+    });
```
Comment (popatre) [inline controllers/users.controller.js:10]: Looks all good here

File: models/comments.model.js [hunk 5]
```diff
@@ -12,28 +12,30 @@ exports.insertComment = async (comment) => {
...
-      status: 404,
-      msg: "User not found",
-    });
-  }
+  // if (await checkUserExists(username)) {
+  //   return Promise.reject({
+  //     status: 404,
+  //     msg: "User not found",
+  //   });
+  // }
```
Comment (popatre) [inline models/comments.model.js:20]: Your commit shouldnt include commented out code.

File: models/comments.model.js [hunk 6]
```diff
@@ -12,28 +12,30 @@ exports.insertComment = async (comment) => {
...
 
-  if (await checkArticleExists(article_id)) {
-    return Promise.reject({
-      status: 404,
-      msg: "Articles not found",
+  // if (await checkArticleExists(article_id)) {
+  //   return Promise.reject({
+  //     status: 404,
+  //     msg: "Articles not found",
+  //   });
```
Comment (popatre) [inline models/comments.model.js:26]: Your commit shouldnt include commented out code. Please remove it from the commit

File: models/comments.model.js [hunk 7]
```diff
@@ -12,28 +12,30 @@ exports.insertComment = async (comment) => {
...
+  return db
+    .query(
+      `INSERT INTO comments (body, author, article_id) VALUES ($1, $2, $3) RETURNING *`,
+      [body, username, article_id]
+    )
+    .then((result) => {
+      return result.rows[0];
+    })
+    .catch((err) => {
+      return { code: err.code, detail: err.detail };
```
Comment (popatre) [inline models/comments.model.js:37]: You should only have one catch, which is in your controller. That should then pass the error to the error handlers. 
Please refactor this to have a single catch at the end of the chain - the chain should end in the controller

File: models/users.model.js [hunk 8]
```diff
@@ -10,4 +10,11 @@ async function checkUserExists(username) {
       return 0;
     });
 }
+
+exports.fetchUsers = () => {
+  return db.query(`SELECT * FROM users`).then((data) => {
+    return data.rows;
+  });
+};
+
```
Comment (popatre) [inline models/users.model.js:19]: Looks all good here

**Review by popatre — CHANGES_REQUESTED** [review]
Hello. It needs a refactor so that your error handlers deal with the errors, not the controller - you've got two catch blocks, which is whats causing you to deal with it in the controller

File: controllers/comments.controller.js [hunk 9]
```diff
@@ -9,16 +9,6 @@ exports.postComment = async (req, resp, next) => {
   };
   insertComment(comment)
     .then((data) => {
```
Comment (popatre) [inline controllers/comments.controller.js:11]: Much cleaner!

File: controllers/errorHandlers.controller.js [hunk 10]
```diff
@@ -1,6 +1,13 @@
 exports.handlePSQL400s = (err, req, resp, next) => {
   if (err.code === "22P02") {
     resp.status(400).send({ msg: "Invalid ID" });
+  }
+  if (err.code === "23503") {
+    const table = err.detail.split(" ").at(-1);
+    if (table === '"users".') {
+      resp.status(404).send({ msg: "User not found" });
+    }
+    resp.status(404).send({ msg: "Article not found" });
```
Comment (popatre) [inline controllers/errorHandlers.controller.js:10]: Excellent stuff!

**Review by popatre — APPROVED** [review]
Spot on, well done 👍

**Pull request merged**